New pages: ShadowRoot.elementFromPoint() (#37115)
mdn/content: files/en-us/web/api/shadowroot/elementsfrompoint/index.md

---
title: "ShadowRoot: elementsFromPoint() method"
short-title: elementsFromPoint()
slug: Web/API/ShadowRoot/elementsFromPoint
page-type: web-api-instance-method
browser-compat: api.ShadowRoot.elementsFromPoint
---

{{APIRef("DOM")}}{{Non-standard_Header}}

> The **`elementsFromPoint()`** method of the {{domxref("ShadowRoot")}} interface returns an array of all the shadow root elements at the specified coordinates (relative to the viewport). The elements are ordered from the element in the topmost layer, to the bottommost element.

Reviewer: ```suggestion
The **`elementsFromPoint()`** method of the {{domxref("ShadowRoot")}} interface returns an array of all the shadow root elements at the specified coordinates (relative to the viewport). The elements are ordered from the topmost element (highest in the display z-order), to the bottommost element.
```

---
title: "ShadowRoot: elementsFromPoint() method"
short-title: elementsFromPoint()
slug: Web/API/ShadowRoot/elementsFromPoint
page-type: web-api-instance-method
browser-compat: api.ShadowRoot.elementsFromPoint
---

{{APIRef("DOM")}}{{Non-standard_Header}}

The **`elementsFromPoint()`** method of the {{domxref("ShadowRoot")}} interface returns an array of all the shadow root elements at the specified coordinates (relative to the viewport). The elements are ordered from the element in the topmost layer, to the bottommost element.

> It operates in a similar way to the {{domxref("ShadowRoot.elementFromPoint")}} method. Some browsers return only the shadow root elements present at that location. Other browsers include elements outside of the {{glossary("shadow DOM")}}, from the shadow DOM element in the topmost layer to the document root node, such as the {{htmlelement("html")}} or {{SVGElement("svg")}} root element. In these browsers, it operates similar to the {{domxref("Document.elementsFromPoint")}} method, but with the ability to cross the [shadow boundary](/en-US/docs/Glossary/Shadow_tree).

Author: ```suggestion
It operates in a similar way to the {{domxref("ShadowRoot.elementFromPoint")}} method. Some browsers return only the shadow root elements present at that location. Other browsers include elements outside of the [shadow DOM](/en-US/docs/Web/API/Web_components/Using_shadow_DOM), from the shadow DOM element in the topmost layer to the document root node, such as the {{htmlelement("html")}} or {{SVGElement("svg")}} root element. In these browsers, it operates similar to the {{domxref("Document.elementsFromPoint")}} method, but with the ability to cross the [shadow boundary](/en-US/docs/Glossary/Shadow_tree).
```

---
title: "ShadowRoot: elementsFromPoint() method"
short-title: elementsFromPoint()
slug: Web/API/ShadowRoot/elementsFromPoint
page-type: web-api-instance-method
browser-compat: api.ShadowRoot.elementsFromPoint
---

{{APIRef("DOM")}}{{Non-standard_Header}}

> The **`elementsFromPoint()`** method of the {{domxref("ShadowRoot")}} interface returns an array of all the shadow root elements
> at the specified coordinates (relative to the viewport). The elements are ordered from the element in the top most layer, to the bottommost element.

Reviewer: ```suggestion
The **`elementsFromPoint()`** method of the {{domxref("ShadowRoot")}} interface returns an array of all the shadow root elements at the specified coordinates (relative to the viewport). The elements are ordered from the element in the top most layer, to the bottommost element.
```

> It operates in a similar way to the {{domxref("ShadowRoot.elementFromPoint")}} method. Some browsers return only the shadow root elements present at that location. Other browsers include elements outside of the {{glossary("shadow DOM")}}, from the shadow DOM element in the topmost layer to the document root node, such as the {{htmlelement("html")}} or {{SVGElement("svg")}} root element. In these browsers, it operates similar to the {{domxref("Document.elementsFromPoint")}} method, but with the ability to cross the [shadow boundary](/en-US/docs/Glossary/Shadow_tree).

Reviewer: `{{glossary("shadow DOM")}}` is a red link. Normally this points to the shadow tree.

> It operates in a similar way to the {{domxref("ShadowRoot.elementFromPoint")}} method. Some browsers return only the shadow root elements present at that location. Other browsers include elements outside of the {{glossary("shadow DOM")}}, from the shadow DOM element in the topmost layer to the document root node, such as the {{htmlelement("html")}} or {{SVGElement("svg")}} root element. In these browsers, it operates similar to the {{domxref("Document.elementsFromPoint")}} method, but with the ability to cross the [shadow boundary](/en-US/docs/Glossary/Shadow_tree).

Reviewer: > Some browsers return only the shadow root elements present at that location. Other browsers include elements outside of

Sounds like a compatibility issue. Should it be in the bcd? Appreciate there is no spec, so there is not "right" way, but there are still behavioural differences.

Author: there's no spec, so not sure who does it right.

Reviewer: Yeah, that sux and is a question for the spec authors - I'd ask against the spec. For now you'd have to note that both are possible.

Reviewer: But there are none. I'm too tired for this.

Reviewer: OK, as there is no spec this would be a pain to do in BCD - you'd have to add a note for each case. So what you have done is pragmatic.

> It operates in a similar way to the {{domxref("ShadowRoot.elementFromPoint", "elementFromPoint()")}} method. Some browsers return only the shadow root elements present at that location. Other browsers include elements outside of the {{glossary("shadow DOM")}}, from the topmost shadow DOM element to the document root node, such as the {{htmlelement("html")}} or {{SVGElement("svg")}} root element. In these browsers, it operates similar to the {{domxref("Document.elementsFromPoint", "elementFromPoint()")}} method, but with the ability to cross the [shadow boundary](/en-US/docs/Glossary/Shadow_tree).

Reviewer 2: This paragraph is confusing because the link text `elementFromPoint()` is the same although the actual linked page is different.

## Syntax

```js-nolint
elementsFromPoint(x, y)
```

### Parameters

- `x`
  - : The horizontal coordinate of a point.
- `y`

>   - : The vertical coordinate of a point.

Reviewer 2: The other page specified the origin for these, but this doesn't.

### Return value

An array of {{domxref('Element')}} objects.

## Examples

```js
let customElem = document.querySelector("my-custom-element");

> let shadow = customElem.shadowRoot;

Reviewer 2: Could be `const`.

> let msg = "";
> let elements = shadow.elementsFromPoint(20, 20);
> elements.forEach((el, i) => {
>   msg += el.localName;
>   if (i < elements.length - 1) {
>     msg += " < ";
>   }
> });

Reviewer 2: ```suggestion
const elements = shadow.elementsFromPoint(20, 20);
const msg = elements.map(el => el.localName).join(" < ");
```

if (msg) {
  console.log(msg);
} else {
  console.log("The custom element had no descendants at x: 20, y: 20.");
}
```

If `<my-custom-element>` is near the top left corner of the viewport, and contains a single `<div>`, the above may return either of the following, depending on the browser implementation:

```plain
div

> div > my-custom-element > body > html

Reviewer 2: ```suggestion
div < my-custom-element < body < html
```
